pkg/config/config.go

```diff
@@ -1,27 +1,179 @@
 package config
 
-type Group struct {
-	GrafanaAdmin bool  `json:"grafanaAdmin,omitempty"`
-	Orgs         []Org `json:"orgs"`
+import (
+	"fmt"
+	"os"
+	"sync"
+
+	"github.com/fsnotify/fsnotify"
+	log "github.com/sirupsen/logrus"
+	"sigs.k8s.io/yaml"
+)
+
+type (
+	Config struct {
+		mu     sync.RWMutex
+		groups GroupsMap
+	}
+
+	// for marshalling purposes
+	config struct {
+		Groups GroupsMap `json:"groups"`
+	}
+
+	GroupsMap map[string]Group
+
+	Group struct {
+		GrafanaAdmin bool  `json:"grafanaAdmin,omitempty"`
+		Orgs         []Org `json:"orgs"`
+	}
+
+	Org struct {
+		ID   int64  `json:"id"`
+		Role string `json:"role"`
+	}
+)
+
+func New() *Config {
+	groupsMap := make(GroupsMap)
+
+	return &Config{
+		groups: groupsMap,
+	}
+}
+
+func NewFromGroupsMap(gm GroupsMap) *Config {
+	return &Config{
+		groups: gm,
+	}
+}
+
+func (c *Config) SetGroup(name string, group Group) {
+	c.mu.Lock()
+	c.groups[name] = group
```

Comment:
This setter needs a nil check on groups to ensure it isn't a nil pointer.

Reply:
sorry don't understand, groups is initialized in New() and NewFromGroupsMap()

Reply:
The groups field is private so you cannot create add it when creating a struct literal but Config is public so the following would panic:

```
	conf := config.Config{}
	fmt.Println(conf.GroupNames())
	conf.SetGroup("thing", config.Group{})
```

Yes, I should use the constructors but nothing enforces that.

Reply:
true, there's no enforcement, so should I create the map in SetGroup if it's nil, no?

Reply:
If you don't want to return an error that would be the most logical option.